man/ssdp-scan.1
.\"                                                              -*- nroff -*-
.\" Copyright (c) 2017-2023  Joachim Wiberg <troglobit@gmail.com>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.a
.Dd Jun 3, 2023
.Dt ssdp-scan 1
.Os
.Sh NAME
.Nm ssdp-scan
.Nd Scan for SSDP capable devices on attached networks
.Sh SYNOPSIS
.Nm
.Op Fl hj
.Op Fl l Ar LEVEL
.Op Fl t Ar SEC
.Op Ar IFACE Op Ar IFACE Op Ar ...
.Sh DESCRIPTION
.Nm
is a tool for scanning for Simple Service Discovery Protocol (SSDP)
capable devices on attached networks.  It sends
.Cm M-SEARCH *
to
.Cm 239.255.255.250
to continuously update a list of SSDP hosts.
.Sh OPTIONS
The following command line options are available:
.Bl -tag -width "-l LEVEL "
.It Fl h
Show program help text.
.It Fl j
JSON output, best used with
.Fl t Ar SEC
option.
.It Fl l Ar LEVEL
Set log level to one of the following, default
.Nm notice :
.Pp
.Bl -tag -width WARNING -compact
.It Cm none
Disable all logging
.It Cm error
Error conditions
.It Cm warning
Warning conditions
.It Cm notice
Normal but significant condition (default)
.It Cm info
Informational
.It Cm debug
Debug-level messages
.El
.Pp
..It Fl t Ar SEC
Timeout and exit after
.Ar SEC
seconds.
.It Op Ar IFACE
By default
.Nm
runs on all IPv4 capable interfaces, except loopback.  If interface
names are given as arguments,
.Nm
will run only on them.
.El
.Sh EXAMPLE
.Bd -literal
$ ssdp-scan -j -t 3
ssdp-scan: Auto-stop in 3 seconds ...
[
  {
    "name": "ReadyNAS Ultra [readynas]",
    "url": "https://192.168.1.171/admin/"
  },
  {
    "name": "readynas: ReadyNAS Ultra 4",
    "url": "http://192.168.1.171:8200/"
  },
  {
    "name": "Kodi (LibreELEC)",
    "url": "http://192.168.1.226:8080/"
  }
]
.Ed
.Sh NOTES
.Nm
is very impatient.  In order to scan for hosts it sends
.Cm M-SEARCH *
messages every other second, which can result in lots of replies, which
in turn causes
.Nm
to retrieve XML files from each replying host.  Be careful.
.Pp
Unless started with the
.Fl t Ar SEC
option,
.Nm
does not terminate on its own.  Instead, it scans continuously until the
user stops it by pressing Ctrl-C.
.Pp
.Nm
does not rely on a local SSDP responder.  It has no dependencies besides
a standard UNIX libc.
.Sh SEE ALSO
.Xr ssdpd 8
.Sh AUTHORS
.Nm
was written by
.An Joachim Wiberg Aq troglobit@gmail.com
to test
.Nm ssdpd
in complex network topologies.  It is modeled on
.Xr mdns-scan 1 .


ChangeLog.md
Change Log
==========

All notable changes to the project are documented in this file.


[v2.1][] - 2023-06-03
---------------------

### Changes
 - Add support for JSON output with `ssdpd-scan -j`, issue #17
 - Dump received data from devices in `ssdp-scan -l debug`
 - Termios support for disabling `^C` output from `ssdp-scan`
 - Update copyright years, LICENSE file hash change
 - Source `/etc/default/ssdpd` from systemd unit file

### Fixes
 - Fix #15: add fallback to `/var/lib/ssdpd/uuid.cache`
 - Fix #16: truncated UUID when reading from cache file


[v2.0][] - 2023-02-19
---------------------

The drop privs release.

### Changes
 - Add support for `-c NAME` to override `ssdpd` UUID cache file
 - Add support for `-d URL` to override the UPnP description URL, a
   single `%s` is supported to be replaced with the interface address
 - Add support for `-m NAME` to override the `manufacturer` in the
   default `description.xml`
 - Add support for `-M URL` to override the `manufacturerURL` in the
   default `description.xml`
 - Add support for `-p URL` to override the `presentationURL` in the
   default `description.xml`, a single `%s` is supported
 - Add support for `-P FILE` to override the default PID file location
   default `description.xml`, a single `%s` is supported
 - Add support for `-u UUID` to use a custom UUID, useful when the
   built-in micro HTTP server is disabled
 - Add support for `-w` to disable built-in micro HTTP server, useful
   when other, more capable, web servers are available.  Make sure to
   have the alternate web server running on port 1901 to serve the file
   `/description.xml`, see also `-d URL` above, which details the
   location of the UPnP description URL
 - Increased debug messages in `ssdp_init()`, for troubleshooting
 - Add support for dropping root privileges after initial setup.  The
   used `ssdp` is searched for first, with fallback to `nobody`
 - Add support for an initial retry interval at startup, in case ssdpd
   is started before any interface has been configured.  Configurable
   retry count using `-R NUM` command line option
 - Add support for an IP address monitor using Linux netlink interfaces
   coming and going at runtime.  Meaning, `-R NUM` is unused on Linux.
   For this to work, `ssdpd` must start as root or have `CAP_NET_RAW`

### Fixes
 - Fix #11: periodic busy loop causing intermittent 100% CPU load
 - Fix invalid `<UDN>uuid:uuid:...</UDN>` in `description.xml`
 - Add `Date:` and `Server:` to HTTP header in micro HTTP server
 - Add support for HTTP HEAD requests to micro HTTP server
 - Handle case when `/etc/os-release` does not have `VERSION_ID` fall
   back to use `VERSION_CODENAME`
 - Don't overwrite CPPFLAGS from the command line
 - Portability fix to `utimensat()` replacement function


[v1.9][] - 2022-10-30
---------------------

The [Dennis Ritchie](https://www.oreilly.com/content/dennis-ritchie-day/) release.

### Changes
 - Add command line options to `sscp-scan`, mostly for testing but may
   be useful for other purposes too
 - Use `$ac_default_prefix` instead of `/usr/local` in configure script
   when expanding paths
 - Update copyright years (affects LICENSE file hash)
 - Add ChangeLog to project
 - Drop Travis-CI in favor of GitHub Actions

### Fixes
 - Fix #6: workaround for OpenVPN /32 default server setup
 - Fix #9: time-of-check vs time-of-use issue with caching of UUID,
   found by Coverity Scan, fixed by Raul Porancea
 - Fix #10: basic instructions for building and starting the daemon


[v1.8][] - 2021-01-23
---------------------

### Changes
 - Use UUID cache directory from configure script, with fall-back to
   operating system specific `/var/lib/misc` or `/var/db`
 - Update man page with info on UUID cache location
 - Add missing systemd unit file
 - Add missing SysV init script

### Fixes
 - Fix avahi-daemon (mDNS) conflict, no conflict with SSDP, caused by
   copy-paste between mdnsd and project and this
 - Fix default install prefix, should be GNU `/usr/local` not `/`.  The
   default for Debian systems is `/usr`
 - Workaround for `--runstatedir` on systems with older autoconf


[v1.7][] - 2020-06-07
---------------------

### Changes
 - Use `/etc/os-release` as base for SSDP server string, distribution
   release information is primarily stored in this file, some Linux
   systems still use `/etc/lsb-release`, on real UNIX systems we can use
   `uname(1)`
 - Update [ssdpd(8)](https://man.troglobit.com/man8/ssdpd.8.html) man page

### Fixes
 - N/A


[v1.6][] - 2020-06-07
---------------------

### Changes
 - Refactor socket handling, enable `SO_REUSEADDR` + `SO_REUSEPORT`
 - Change default behavior, must now use `-n` to run in foreground
 - Add `ssdp-scan` tool, similar to `mdns-scan`
 - Add systemd unit file
 - Initial Debian/Ubuntu packaging

### Fixes
 - Fix #1: CVE-2019-14323


[v1.5][] - 2017-09-23
---------------------

### Changes
 - New default SSDP notify interval, 300 sec. (Was 600 sec)
 - Validation of refresh and notify intervals, the notify interval must
   not be longer than half the cache timeout
   - Notify: 30-900 sec
   - Refresh: 5-1800 sec

### Fixes
 - Fix regression in v1.3, allow running without interface filtering
 - The new interface filtering feature of v1.4 managed to filter out all
   interfaces if none were given on the command line.
 - Fix CPU overload problem, caused by invalid timer comparison


[v1.4][] - 2017-05-16
---------------------

Minor bug fix release.

### Fixes
 - Fix invalid argument to `accept()` in web server


[v1.3][] - 2017-05-16
---------------------

### Changes
 - Massive refactor
 - Support SSDP per interface and multiple addresses per interface
 - Use `SOCK_DGRAM` instead of `SOCK_RAW`, no longer need to run as root
 - New refresh timer, runs independently of NOTIFY timer, checks for new addresses


[v1.2][] - 2017-05-16
---------------------

Announces itself as an InternetGatewayDevice, works with Windows, serves
XML description.

### Changes

src/ssdp-scan.c
/* SSDP scanner
 *
 * Copyright (c) 2017-2023  Joachim Wiberg <troglobit@gmail.com>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.a
 */

#include "ssdp.h"
#include "queue.h"

#include <err.h>
#ifdef HAVE_TERMIOS_H
# include <termios.h>
#endif


struct host {
	LIST_ENTRY(host) link;

	char *name;
	char *url;
};

LIST_HEAD(, host) hl = LIST_HEAD_INITIALIZER();

volatile sig_atomic_t running = 1;
#ifdef HAVE_TERMIOS_H
struct termios term;
#endif
static int atty;
static int json;
static int once;

extern FILE *uget(char *url);


static void hidecursor(void)
{
#ifdef HAVE_TERMIOS_H
	struct termios c;
#endif

	if (!atty)
		return;

#ifdef HAVE_TERMIOS_H
	if (tcgetattr(0, &term)) {
		atty = 0;
		return;
	}
	c = term;
	c.c_lflag &= ~ECHO;
	tcsetattr(0, TCSANOW, &c);
#endif
	fputs("\e[?25l", stdout);
}

static void showcursor(void)
{
	if (!atty)
		return;

#ifdef HAVE_TERMIOS_H
	tcsetattr(0, TCSANOW, &term);
#endif
	fputs("\e[?25h", stdout);
}

static int host(char *name, char *url)
{
	struct host *h;

	if (!name || !url)
		return 1;

	LIST_FOREACH(h, &hl, link) {
		if (strcmp(h->name, name))
			continue;
		if (strcmp(h->url, url))
			continue;

		return 1;
	}

	h = malloc(sizeof(*h));
	h->name = strdup(name);
	h->url  = strdup(url);

	LIST_INSERT_HEAD(&hl, h, link);

	return 0;
}

static void progress(void)
{
//	const char *style = ".oOOo.";
	const char *style = "\\-/|";
//	const char *style = ">v<^";
	static unsigned int i = 0;
	size_t num = 4;

	if (!atty || json)
		return;

//	printf("%u %% %zd = %lu (%u / %zd = %lu)\n", i, num, i % num, i, num, i / num);

	putchar(style[i++ % num]);
	printf("\b");
	fflush(stdout);
}

static void ssdp_scan(struct ifsock *ifs, int arg)
{
	struct sockaddr_in sin;
	char buf[200];
	ssize_t num;
	int len;

	(void)arg;
	if (!ifs || ifs->sd == -1)
		return;

	memset(buf, 0, sizeof(buf));
	len = snprintf(buf, sizeof(buf), "M-SEARCH * HTTP/1.1\r\n"
		       "Host: %s:%d\r\n"
		       "ST: %s\r\n"
		       "Man: \"ssdp:discover\"\r\n"
		       "MX: 3\r\n"
		       "\r\n",
		       MC_SSDP_GROUP, MC_SSDP_PORT,
		       "upnp:rootdevice");

	memset(&sin, 0, sizeof(sin));
	sin.sin_family      = AF_INET;
	sin.sin_port        = htons(MC_SSDP_PORT);
	sin.sin_addr.s_addr = inet_addr(MC_SSDP_GROUP);

	num = sendto(ifs->sd, buf, len, 0, (struct sockaddr *)&sin, sizeof(sin));
	if (num < 0)
		warn("Failed sending SSDP M-SEARCH");
}

static char *find(char *buf, const char *search)
{
	char *ptr;

	ptr = strcasestr(buf, search);
	if (!ptr)
		return NULL;

	ptr += strlen(search);
	while (*ptr && *ptr == ' ')
		ptr++;

	return ptr;
}

static char *trim(char *ptr)
{
	char *end = ptr;

	while (*end && *end != '\r' && *end != '\n')
		end++;
	*end = 0;

	return ptr;
}

static int xml(char *buf, char *tagn, char *val, size_t vlen)
{
	char *ptr, *end;
	char tag[20];
	size_t len;

	len = snprintf(tag, sizeof(tag), "<%s>", tagn);
	ptr = strstr(buf, tag);
	if (!ptr)
		return 0;
	ptr += len;

	len = snprintf(tag, sizeof(tag), "</%s>", tagn);
	end = strstr(buf, tag);
	if (end)
		*end = 0;

	strlcpy(val, ptr, vlen);

	return 1;
}

static void parse(FILE *fp, char **name, char **url)
{
	static char uri[80];
	static char nm[80];
	char buf[512];

	memset(uri, 0, sizeof(uri));
	memset(nm, 0, sizeof(nm));

	while (fgets(buf, sizeof(buf), fp)) {
		if (!nm[0] && xml(buf, "friendlyName", nm, sizeof(nm)))
			*name = nm;
		if (!uri[0] && xml(buf, "presentationURL", uri, sizeof(uri)))
			*url = uri;

		if (*name && *url)
			break;
	}
}

static void print(char *name, char *url)
{
	if (json)
		printf("%s  {\n"
		       "    \"name\": \"%s\",\n"
		       "    \"url\": \"%s\"\n"
		       "  }", once++ ? ",\n" : "", name, url);
	else
		printf("\r+ %-40s  %s\n", name, url);
}

static void printsrv(char *srv, char *loc)
{
	char *name = NULL, *url = NULL;
	char *copy;
	FILE *fp;

	/* Save copy in case of short/empty presentationURL */
	copy = strdup(loc);
	if (!copy)
		return;

	if (strncmp(loc, "http", 4))
		goto fallback;

	fp = uget(loc);
	if (!fp) {
	fallback:
		print(trim(srv), trim(loc));
		free(copy);
		return;
	}

	parse(fp, &name, &url);
	fclose(fp);

	if (!name || !url) {
		free(copy);
		return;
	}

	if (url && url[0] == '/') {
		char *ptr;

		copy = realloc(copy, strlen(copy) + strlen(url));
		if (!copy)
			return;

		ptr = strstr(copy, "://");
		if (ptr)
			ptr += 3;
		else
			ptr = copy;

		ptr = strchr(ptr, '/');
		if (ptr)
			strcpy(ptr, url);
		url = copy;
	}

	if (!host(name, url))
		print(name, url);

	free(copy);
}

static void ssdp_read(int sd)
{
	char buf[512];
	ssize_t len;
	char *loc;
	char *srv;

	memset(buf, 0, sizeof(buf));
	len = recv(sd, buf, sizeof(buf) - 1, 0);
	if (len <= 0)
		return;

	buf[len] = 0;
	logit(LOG_DEBUG, "%s(): buf[%s]", __func__, buf);

	if (strstr(buf, "M-SEARCH *"))
		return;

	loc = find(buf, "Location:");
	srv = find(buf, "Server:");

	if (!loc || !srv)
		return;

	trim(loc);
	trim(srv);

	printsrv(srv, loc);
}

static void bye(int signo)
{
	(void)signo;
	running = 0;
}

static int usage(int code)
{
	printf("Usage: ssdp-scan [-hj] [-l LEVEL] [-t SEC] [IFACE [IFACE ...]]\n"
	       "\n"
	       "    -h        This help text\n"
	       "    -j        JSON output, best used with -t SEC option"
	       "    -l LEVEL  Set log level: none, err, notice (default), info, debug\n"
	       "    -t SEC    Timeout and exit after SEC seconds.\n"
	       "\n"
	       "Bug report address : %s\n", PACKAGE_BUGREPORT);
#ifdef PACKAGE_URL
        printf("Project homepage   : %s\n", PACKAGE_URL);
#endif

	return code;
}

int main(int argc, char *argv[])
{
	int throttle = 1;
	int timeout = 0;
	int c;

	log_level = LOG_NOTICE;
	atty = isatty(STDOUT_FILENO);

	while ((c = getopt(argc, argv, "hjl:t:")) != EOF) {
		switch (c) {
		case 'h':
			return usage(0);

		case 'j':
			json = 1;
			break;

		case 'l':
			log_level = log_str2lvl(optarg);
			if (-1 == log_level)
				return usage(1);
			break;

		case 't':
			timeout = atoi(optarg);
			break;

		default:
			return usage(1);
		}
	}

	signal(SIGINT, bye);
	signal(SIGALRM, bye);

	log_init(0);

	if (timeout > 0) {
		warnx("Auto-stop in %d seconds ...", timeout);
		alarm(timeout);
	}

	if (ssdp_init(1, 0, &argv[optind], argc - optind, ssdp_read) < 1)
		return 1;

	hidecursor();
	progress();

	if (json)
		printf("[\n");

	ssdp_foreach(ssdp_scan, 1);

	while (running) {
		progress();

		switch (ssdp_poll(100)) {
		case -1:
			if (errno == EINTR)
				break;

			err(1, "Unrecoverable error");
			break;

		case 0:
			if (!(throttle++ % 20))
				ssdp_foreach(ssdp_scan, 0);
			break;

		default:
			break;
		}
	}

	if (json)
		printf("%s]\n", once ? "\n" : "");

	showcursor();
	log_exit();

	return ssdp_exit();
}

/**
 * Local Variables:
 *  indent-tabs-mode: t
 *  c-file-style: "linux"
 * End:
 */
